### ISSUE TITLE
AP: alternative opt in mechanisms

### ISSUE THREAD

**Issue by sysadminpower2019** [snarfed/bridgy-fed#1051]
Hello,

Castopod has native AP support but there is no way to add or follow people directly on the admin panel. So I am not sure how you could / would integrate castopod to bridgy-fed

**Comment by snarfed** [comment]
Interesting. Can you DM? If you DM _yes_ to @bsky.brid.gy@bsky.brid.gy, that also enables the bridge.

**Comment by sysadminpower2019** [comment]
It doesn't seem possible either. It seems to be a sorta one way app. 
This is why i thought the process would something like submitting the fediverse site to the bridge site and conforming maybe via DNS changes

**Comment by snarfed** [comment]
DNS! That would be...doable, but ambitious.

Ideally, I want opt-in mechanism(s) to be:
* easy to use
* verify that you own the account
* private
* generic, ie not specific to individual servers like Castopod

So far, following and DMing satisfy these criteria. I know they're not universally supported, though, so I'm open to other ideas!

**Comment by qazmlp** [comment]
There are I think only two mechanisms that are *near* universal for fediverse actors: The profile `"summary"` (bio) and posting statuses (though definitely not everything is able to post `Note`s).

---

The profile bio flow is easier (and I'm just using specific tokens as example, another format may be better):

- Ask the user to temporarily include some text like `<Bridgy Fed: yes>` in their bio.
- Let them submit their @@-handle or AP `"id"` in a form on the Bridgy Fed website.
- Check the profile bio, and do one of the following:

  |previous state|bio token|action|response to user|
  |-|-|-|-|
  |opted-out|neither|none|instruct to add `<Bridgy Fed: yes>` to bio, wait a minute, and retry|
  |either|`<Bridgy Fed: no>`<br>(or both)|opt-out|confirm `<Bridgy Fed: no>` and opt-out,<br>note that bridged content may disappear with a delay|
  |opted-in|not `<Bridgy Fed: no>`|profile refresh|confirm previous opt-in,<br>present bridged handle,<br>instruct to add `<Bridgy Fed: no>` and resubmit in order to opt 'back out' |
  |opted-out|`<Bridgy Fed: yes>` and not `<Bridgy Fed: no>`|opt-in|confirm previous opt-in,<br>present bridged handle,<br>instruct to add `<Bridgy Fed: no>` and resubmit in order to opt 'back out'|

…or at least that's how I would do this, plus various error messages for other kinds of failures like not being able to fetch the actor.

Consider also opting out accounts if you see `<Bridgy Fed: no>` during incidental profile updates, since I think the intention is clear but the user may not remember to use the form.

Notably, bird.makeup accounts can't use this flow because it doesn't bridge the bio (but bird.makeup bridges without consent in the first place, so it's a bit ehhh).

---

For the status flow, you can do essentially the same thing, but (also) accept a link to a status in the same form. (If you don't find tokens in the text, you probably should fall back to the author's bio automatically.)

As added complication, this would require saving the timestamp, so that no-one else can opt an account back in or out using an older status that wasn't deleted. I'd still validate the timestampe with `>=` in case the form accidentally gets submitted twice.

**Comment by sysadminpower2019** [comment]
Would it not be possible / simpler to have someone insert a txt record in the domain and then read this txt record and be done with it

**Comment by snarfed** [comment]
That would only work for an entire instance, not for individual users, and we'd still need a way to trigger Bridgy Fed to _look_ at that instance and the DNS record anyway.

**Comment by pcottle** [comment]
I think having a way to opt in via text in bio would be great! It'd empower individual users but not require any additional code to be written in the implementation of the server.

**Comment by Tamschi** [comment]
This would also enable some level of Threads-support. See https://github.com/snarfed/bridgy-fed/issues/1210#issuecomment-2380637922

**Comment by snarfed** [comment]
I think https://github.com/snarfed/bridgy-fed/issues/1680 has now subsumed this...?

**Comment by snarfed** [comment]
Duplicate of #1680

**Comment by snarfed** [comment]
Good news is, this is now a higher priority for us!

**Comment by lalomartins** [comment]
Hi,

This is closed as duplicate of #1680, but the solution for #1680 doesn't meet this use case. Apps that federate to ap like castopod, peertube, owncast, and writefreely don't necessarily support oauth login.

I'd still love to be able to bridge my writefreely blog to bluesky somehow.

**Comment by snarfed** [comment]
@lalomartins good point! Those all can't do outbound follows _or_ DMs _or_ OAuth? If not, then agreed, we can reopen this.

**Comment by lalomartins** [comment]
I haven't checked all the other apps listed here, but the one I'm using (writefreely) can't, it only publishes. I assume the others are the same.

The vibe I'm getting is that publishing to AP is the new RSS, so it would be a good use case to support IMO. As for how… I'll leave that challenge to the team 😅 maybe some sort of “well known URL” file with json/yaml/whatever settings might work for self-hosted, but a solution that works for hosted too would be even better. (Then again I know next to nothing about the AP protocol itself)

**Comment by snarfed** [comment]
Reconsidering this. We now have at least four ways for people on the fediverse to bridge themselves: follow, DM, OAuth login on https://fed.brid.gy/ , and [instance level](https://fed.brid.gy/docs#enable-api). That's a lot. If apps like Castopod etc can't do _any_ of those, that's disappointing, but realistically I doubt we'd add and maintain a fifth method any time soon.

**Comment by lalomartins** [comment]
Please consider the use cases before disposing of this particular bath water.

(By the way,  “instance level” is not really a fourth way, it's just a follow as far as your code is concerned, at least judging from the link in your comment.)

We're increasingly seeing apps use AP as a **publishing** protocol. And honestly, fair enough? It's called ActivityPub after all, not ActivityComm or ActivitySoc. These apps have no legitimate use case for implementing things like following or DMs. What does it even mean for a blog or podcast or video channel to follow someone? The apps just generate posts and send them out there into the fediverse.

I could understand the reluctance if we were talking one single odd corner case, but this thread has so far conjured 4 examples (from 7 participants). The way these things usually go, that means there are probably at least 10 or so. I would say it looks like a trend… and something that can easily make Bridgy even more useful.

**Comment by snarfed** [comment]
I'm absolutely with you on the use case! Ideally we want all of the remaining apps here to be able to use the bridge too. My reluctance is about the increasing amount of complexity we take on for each additional opt in mechanism.

**Comment by lalomartins** [comment]
I hear you. I'm a tech lead for my day job, and legacy code, cognitive overload, and overdesign take a good chunk of my time.

How about flipping things around? Instead of framing this as “add another opt-in method”, why don't we think of it as, “support AP as a publishing protocol for web”? I can opt in the way I would normally opt in for a website, but then the site publishes AP instead of Atom or Microformats. Would that add less complexity?

**Comment by snarfed** [comment]
I appreciate the creativity! The problem is that anyone can bridge a web site, but #835 etc stopped us from doing that for fediverse accounts.

**Comment by lalomartins** [comment]
Ah. Tough one. 🤔

Still, code-wise it seems like it might be the best solution? It could be argued that “anyone can bridge a web site” is probably not the desired behavior anyway, for similar reasons as described in #835. Having a layer of control via a well-known URL or a domain record or something would be the natural way around that, and coincidentally those are solutions proposed for this ticket. So it looks to me even more like, conceptually, that's what we're talking about: a website that happens to publish via AP, not a regular fediverse account.

Is there an easy way to distinguish between those two things? A quick browse of the specs suggests that Actor type might be useful, but I don't know how consistently that's implemented. (Yeah, a quick check on [writefreely source code](https://github.com/writeas/web-core/blob/cf9f5db56bbc1277ace87be0ea116e594f4f93ba/activitystreams/person.go#L21) says they use type Person.) Stretching it even a little more, if `following` is empty maybe it has no following capabilities? IDK. Would be nice if AP had a “capabilities” concept I guess.

**Comment by snarfed** [comment]
Hmm. For services that are enough like web sites to have RSS feeds and their own (sub)domains, like write.as, you can bridge them _as_ web sites, directly. No need to use AP as an intermediary.

For user-level accounts like PeerTube on servers that someone else runs, users won't have access to either DNS or server-wide paths like `.well-known`, so we wouldn't be able to use those.

**Comment by lalomartins** [comment]
Yeah. Touché. 🤔

**Comment by lalomartins** [comment]
Although I should add, it's not quite “using AP as an intermediary”. There are actual benefits. We get follow counts, like counts, etc, a writefreely (or write.as) blog can even be @-mentioned. I'm not sure how much of that would carry through the bridge, but I for one would rather bridge via AP than RSS, if it worked.

**Comment by snarfed** [comment]
Web bridging does all of those things too, and more! Via webmention + microformats. https://fed.brid.gy/docs#web-get-started
